hazelcast/reactor.py

```diff
@@ -101,33 +258,81 @@ def shutdown(self):
             self._thread.join()
 
         for connection in list(self._map.values()):
+            if connection is self.waker:
+                continue
+
             try:
                 connection.close(None, HazelcastError("Client is shutting down"))
             except OSError as connection:
                 if connection.args[0] == socket.EBADF:
                     pass
                 else:
                     raise
+
+        self.waker.close()
         self._map.clear()
 
-    def connection_factory(self, connection_manager, connection_id, address, network_config, message_callback):
-        return AsyncoreConnection(self._map, connection_manager, connection_id,
-                                  address, network_config, message_callback)
 
-    def _cleanup_all_timers(self):
-        timers = self._timers
-        new_timers = self._new_timers
+class _BusyWaitLoop(_AbstractLoop):
+    def check_loop(self):
+        pass
 
-        while timers:
-            _, timer = timers.pop()
-            timer.timer_ended_cb()
+    def run_loop(self):
+        if not self._map:
+            time.sleep(0.005)
+        asyncore.loop(timeout=0.01, use_poll=True, map=self._map, count=1)
 
-        # Although it is not the case with the current code base,
-        # the timers ended above may add new timers. So, the order
-        # is important.
-        while new_timers:
-            _, timer = new_timers.popleft()
-            timer.timer_ended_cb()
+    def wake_loop(self):
+        pass
+
+    def shutdown(self):
+        if not self._is_live:
+            return
+
+        self._is_live = False
+
+        if self._thread is not threading.current_thread():
+            self._thread.join()
+
+        for connection in list(self._map.values()):
+            try:
+                connection.close(None, HazelcastError("Client is shutting down"))
+            except OSError as connection:
+                if connection.args[0] == socket.EBADF:
+                    pass
+                else:
+                    raise
+
+        self._map.clear()
+
+
+class AsyncoreReactor(object):
+    def __init__(self):
+        self.map = {}
+        loop = None
+        try:
+            loop = _WakeableLoop(self.map)
+            loop.check_loop()
+        except:
+            if loop:
+                loop.shutdown()
+                loop = _BusyWaitLoop(self.map)
```

Comment:
nit: it might be a good idea to log a warn message here, as this implementation may be sub-optimal in certain scenarios.